lib/services/role.go

```diff
@@ -1887,69 +1600,174 @@ func (m *DatabaseNameMatcher) String() string {
 	return fmt.Sprintf("DatabaseNameMatcher(Name=%v)", m.Name)
 }
 
-// CheckAccessToDatabase checks if this role set has access to a particular database.
-//
-// The checker always checks the server namespace, other matchers are supplied
-// by the caller.
-func (set RoleSet) CheckAccessToDatabase(server types.Database, mfa AccessMFAParams, matchers ...RoleMatcher) error {
+type loginMatcher struct {
+	login string
+}
+
+// NewLoginMatcher creates a RoleMatcher that checks whether the role's logins
+// match the specified condition.
+func NewLoginMatcher(login string) RoleMatcher {
+	return &loginMatcher{login: login}
+}
+
+func (l *loginMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
+	logins := role.GetLogins(typ)
+	for _, login := range logins {
+		if l.login == login {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// AccessCheckable is the subset of types.Resource required for the RBAC checks.
+type AccessCheckable interface {
+	GetKind() string
+	GetName() string
+	GetMetadata() types.Metadata
+	GetAllLabels() map[string]string
+}
+
+func rbacDebugLogger() (debugEnabled bool, debugf func(format string, args ...interface{})) {
+	isDebugEnabled := log.IsLevelEnabled(log.DebugLevel)
 	log := log.WithField(trace.Component, teleport.ComponentRBAC)
+	return isDebugEnabled, func(format string, args ...interface{}) {
+		if isDebugEnabled {
+			log.Debugf(format, args...)
+		}
+	}
+}
+
+// CheckAccess checks if this role set has access to a particular resource,
+// optionally matching the resource's labels.
+func (set RoleSet) CheckAccess(r AccessCheckable, mfa AccessMFAParams, matchLabels bool, matchers ...RoleMatcher) error {
+	// Note: logging in this function only happens in debug mode. This is because
+	// adding logging to this function (which is called on every resource returned
+	// by the backend) can slow down this function by 50x for large clusters!
+	isDebugEnabled, debugf := rbacDebugLogger()
+
 	if mfa.AlwaysRequired && !mfa.Verified {
-		log.Debugf("Access to database %q denied, cluster requires per-session MFA", server.GetName())
+		debugf("Access to %v %q denied, cluster requires per-session MFA", r.GetKind(), r.GetName())
 		return ErrSessionMFARequired
 	}
+
+	namespace := types.ProcessNamespace(r.GetMetadata().Namespace)
+	allLabels := r.GetAllLabels()
+
+	var getRoleLabels func(types.Role, types.RoleConditionType) types.Labels
+	if matchLabels {
+		switch r.GetKind() {
+		case types.KindDatabase:
+			getRoleLabels = types.Role.GetDatabaseLabels
+		case types.KindApp:
+			getRoleLabels = types.Role.GetAppLabels
+		case types.KindNode:
+			getRoleLabels = types.Role.GetNodeLabels
+		case kindKubeClusterRBAC:
+			getRoleLabels = types.Role.GetKubernetesLabels
+		default:
+			return trace.BadParameter("cannot match labels for kind %v", r.GetKind())
+		}
+	}
```

Comment:
If we have callers pass their respective label matchers, this won't be needed.

Reply:
That's how it was originally. But it was then possible to call incorrectly.  For example, you could make a mistake like the following and perform checks against a role's database labels for an app.

```
var app types.App
CheckAccess(app, ..., types.Role.GetDatabaseLabels)
```

Reply:
I mean, you can do the same now as well if you pass other incorrect matchers? Not sure how label matcher is different. It's just very weird from the function contract perspective that we seem to be unifying the checkers and matchers but labels are for some reason treated differently.